go/toolchain/toolchains.bzl

```diff
@@ -3,81 +3,48 @@ load('//go/private:go_toolchain.bzl', 'external_linker', 'go_toolchain')
 DEFAULT_VERSION = "1.9"
 
 def _generate_toolchains():
-  # The full set of allowed os and arch combinations for the go toolchain
-  # This is the set of targets allowed, of which the set of hosts is a strict subset
-  android_arm = struct(os="android", arch="arm")
-  darwin_386 = struct(os="darwin", arch="386")
-  darwin_amd64 = struct(os="darwin", arch="amd64")
-  darwin_arm = struct(os="darwin", arch="arm")
-  darwin_arm64 = struct(os="darwin", arch="arm64")
-  dragonfly_amd64 = struct(os="dragonfly", arch="amd64")
-  freebsd_386 = struct(os="freebsd", arch="386")
-  freebsd_amd64 = struct(os="freebsd", arch="amd64")
-  freebsd_arm = struct(os="freebsd", arch="arm")
-  linux_386 = struct(os="linux", arch="386")
-  linux_amd64 = struct(os="linux", arch="amd64")
-  linux_arm = struct(os="linux", arch="arm")
-  linux_arm64 = struct(os="linux", arch="arm64")
-  linux_ppc64 = struct(os="linux", arch="ppc64")
-  linux_ppc64le = struct(os="linux", arch="ppc64le")
-  linux_mips = struct(os="linux", arch="mips")
-  linux_mipsle = struct(os="linux", arch="mipsle")
-  linux_mips64 = struct(os="linux", arch="mips64")
-  linux_mips64le = struct(os="linux", arch="mips64le")
-  netbsd_386 = struct(os="netbsd", arch="386")
-  netbsd_amd64 = struct(os="netbsd", arch="amd64")
-  netbsd_arm = struct(os="netbsd", arch="arm")
-  openbsd_386 = struct(os="openbsd", arch="386")
-  openbsd_amd64 = struct(os="openbsd", arch="amd64")
-  openbsd_arm = struct(os="openbsd", arch="arm")
-  plan9_386 = struct(os="plan9", arch="386")
-  plan9_amd64 = struct(os="plan9", arch="amd64")
-  solaris_amd64 = struct(os="solaris", arch="amd64")
-  windows_386 = struct(os="windows", arch="386")
-  windows_amd64 = struct(os="windows", arch="amd64")
-  
   # The set of acceptable hosts for each of the go versions, this is essentially the
   # set of sdk's we know how to fetch
   versions = [
       struct(
           name = "host",
           sdk = "@go_host_sdk",
-          hosts = [darwin_amd64, linux_386, linux_amd64, windows_386, windows_amd64, freebsd_386, freebsd_amd64],
+          hosts = ["darwin_amd64", "linux_386", "linux_amd64", "windows_386", "windows_amd64", "freebsd_386", "freebsd_amd64"],
```

Comment:
I was curious why we don't have the same set of hosts for all versions. I guess this is derived from `_sdk_repositories` in go/private/repositories.bzl. Could we build this list automatically from that table?

Reply:
Not all versions have all host operating systems, new ones arrive and old ones stop being shipped, so it can't be just one list.
I did not bother adding hosts for old versions, I only added the ones we always supported, but I add all hosts that we ship every time we add a new version.